src/cli/cli_uart.cpp

```diff
@@ -233,6 +234,7 @@ int Uart::Output(const char *aBuf, uint16_t aBufLength)
     }
 
     Send();
+    OT_CLI_UART_OUTPUT_UNLOCK();
```

Comment:
I think we need to protect the `Send()` call here, right?
(or do we need to protect the whole `Output()` call? Do we allow `Output()` to be called from other OS tasks/contexts? ) 

`Uart::Send()` itself does call into platform code 
```
void Uart::Send(void)
{
    VerifyOrExit(mSendLength == 0);

    if (mTxLength > kTxBufferSize - mTxHead)
    {
        mSendLength = kTxBufferSize - mTxHead;
    }
    else
    {
        mSendLength = mTxLength;
    }

    if (mSendLength > 0)
    {
#if OPENTHREAD_CONFIG_ENABLE_DEBUG_UART
        /* duplicate the output to the debug uart */
        otPlatDebugUart_write_bytes(reinterpret_cast<uint8_t *>(mTxBuffer + mTxHead), mSendLength);
#endif
        otPlatUartSend(reinterpret_cast<uint8_t *>(mTxBuffer + mTxHead), mSendLength);
    }

exit:
    return;
}
```
If we need to protect the `Send()` only, then how about adding the lock in platform code in `otPlatUartSend()` (or maybe in `otPlatDebugUart_write_bytes()` if being used).

Reply:
Can you please help explain why having a lock in `otPlatUartSend` won't work for NCP devices? Thanks.

Reply:
Some ncp devices use uart to send spinel messages. If other threads call `otPlatUartSend`, spinel message will get corrupted.

Reply:
Correct. But why the lock cannot be moved into `otPlatUartSend `?
We cannot and should not use simple debug UART log on NCP mode with UART (lock or not lock). But having a lock in platform code should not harm the NCP use-case.

I strongly believe/prefer that OpenThread would not be  involved in any thread/task/OS management (locking and unlocking) and have platform code manage this. In general the OT platform APIs are intended for use by OT only. Everything related to OS should be taken care by the platform APIs themselves.

Reply:
Got it. Thanks. Let's go for the lock in CLI...